### ISSUE TITLE
Is there a servlet -> core adapter?

### ISSUE THREAD

**Issue by cowwoc** [eclipse/jetty.project#11467]
**Jetty Version**
12.0.5

**Jetty Environment**
core

**Question**
I am using jetty-core for my application, but in trying to integrate auth0 support I ran into a problem where they require an instance of `HttpServletRequest`, `HttpServletResponse`.

I don't want to use servlets for the rest of my application. Is there an existing class that implements `HttpServletRequest`, `HttpServletResponse` and translates calls to a core `Request`, `Response`?

**Comment by cowwoc** [comment]
I ran across ServletApiRequest which does what I want, but it's only for ee10. Auth0 depends on ee8. Is there an equivalent class that I can use?

**Comment by gregw** [comment]
@cowwoc have a look at #11451, which extends the availability of ServletCoreRequest to ee8 & ee9.

**Comment by gregw** [comment]
@cowwoc ah those classes are package protected.   Let me see if I can make them public....

**Comment by cowwoc** [comment]
@gregw Even if the classes in question were public, how can I use them to wrap a core `Request` in a `HttpServletRequest`? 

These classes seem to do the opposite: take in an existing `HttpServletRequest` and convert its calls to a new `Request` object.

I want to pass in an existing `Request` object and get back a `HttpServletRequest`. I would then pass the `HttpServletRequest` into auth0's library and its call would modify the underlying `Request` which the rest of my application would use.

**Comment by gregw** [comment]
@cowwoc Ah the other direction!!!!  We do have that in all the environments, but they are full conversions, so they need the support of the context and the ServletHandler (and its ServletChannel) to run the full servlet state machine (startAsync etc.)

It sounds like you want a minimal wrapper just to handle headers and bodies.    That should be trivial enough to do, given that we already have the ability to turn Content into streams easily.  We have no need of those classes, but I can see they are utility enough so that is somebody (hint) provided a PR with them, then we'd include them and maintain them.

A lot of the methods (e.g. `startAsync()`) would have to throw `UnsupportedOperationException`, but there should be enough implementable methods that these classes would be useful

**Comment by cowwoc** [comment]
@gregw Sounds good. I'll try to throw something together and get back to you. Thanks.

**Comment by joakime** [comment]
@cowwoc keep in mind that these new classes would need to live in one of the environments, not core itself.
We don't want to introduce a Servlet dependency in `/jetty-core/`

**Comment by cowwoc** [comment]
I agree that these classes would need to live in one of the EE environments. Here is what I've got so far... I didn't need to implement any of the other methods for auth0 to work:

```
/**
 * Converts the servlet API to Jetty's core API.
 */
public final class JettyServletRequest implements HttpServletRequest
{
	private final Request request;
	private final Response response;

	/**
	 * @param request  the server request
	 * @param response the server response
	 * @throws NullPointerException if any of the arguments are null
	 */
	public JettyServletRequest(Request request, Response response)
	{
		requireThat(request, "request").isNotNull();
		requireThat(response, "response").isNotNull();
		this.request = request;
		this.response = response;
	}

	@Override
	public String getAuthType()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Cookie[] getCookies()
	{
		List<HttpCookie> cookies = Request.getCookies(request);
		List<Cookie> servletCookies = new ArrayList<>();
		for (HttpCookie cookie : cookies)
			servletCookies.add(from(cookie));
		return servletCookies.toArray(new Cookie[0]);
	}

	private static Cookie from(HttpCookie cookie)
	{
		Cookie servletCookie = new Cookie(cookie.getName(), cookie.getValue());
		String comment = cookie.getComment();
		if (comment != null)
			servletCookie.setComment(comment);
		String domain = cookie.getDomain();
		if (domain != null)
			servletCookie.setDomain(domain);
		servletCookie.setHttpOnly(cookie.isHttpOnly());
		servletCookie.setMaxAge(Math.clamp(cookie.getMaxAge(), Integer.MIN_VALUE, Integer.MAX_VALUE));
		String path = cookie.getPath();
		if (path != null)
			servletCookie.setPath(path);
		servletCookie.setSecure(cookie.isSecure());
		servletCookie.setVersion(cookie.getVersion());
		return servletCookie;
	}

	@Override
	public long getDateHeader(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getHeader(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Enumeration<String> getHeaders(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Enumeration<String> getHeaderNames()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getIntHeader(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getMethod()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getPathInfo()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getPathTranslated()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getContextPath()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getQueryString()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getRemoteUser()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isUserInRole(String role)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Principal getUserPrincipal()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getRequestedSessionId()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getRequestURI()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public StringBuffer getRequestURL()
	{
		return new StringBuffer(Uris.withoutQueryStringOrFragment(request.getHttpURI().toURI()).toString());
	}

	@Override
	public String getServletPath()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public HttpSession getSession(boolean create)
	{
		return new JettySession(request, response);
	}

	@Override
	public HttpSession getSession()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String changeSessionId()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isRequestedSessionIdValid()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isRequestedSessionIdFromCookie()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isRequestedSessionIdFromURL()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	@Deprecated
	public boolean isRequestedSessionIdFromUrl()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean authenticate(HttpServletResponse response) throws IOException, ServletException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void login(String username, String password) throws ServletException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void logout() throws ServletException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Collection<Part> getParts() throws IOException, ServletException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Part getPart(String name) throws IOException, ServletException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public <T extends HttpUpgradeHandler> T upgrade(Class<T> handlerClass) throws IOException, ServletException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Object getAttribute(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Enumeration<String> getAttributeNames()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getCharacterEncoding()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setCharacterEncoding(String env) throws UnsupportedEncodingException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getContentLength()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public long getContentLengthLong()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getContentType()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public ServletInputStream getInputStream() throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getParameter(String name)
	{
		return getQueryParameters().getValue(name);
	}

	private Fields getQueryParameters()
	{
		String queryParameters = request.getHttpURI().getQuery();
		if (queryParameters.isBlank())
			return Fields.EMPTY;
		Fields fields = new Fields(true);
		UrlEncoded.decodeUtf8To(queryParameters, fields);
		return fields;
	}

	@Override
	public Enumeration<String> getParameterNames()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String[] getParameterValues(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Map<String, String[]> getParameterMap()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getProtocol()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getScheme()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getServerName()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getServerPort()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public BufferedReader getReader() throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getRemoteAddr()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getRemoteHost()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setAttribute(String name, Object o)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void removeAttribute(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Locale getLocale()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Enumeration<Locale> getLocales()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isSecure()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public RequestDispatcher getRequestDispatcher(String path)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	@Deprecated
	public String getRealPath(String path)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getRemotePort()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getLocalName()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getLocalAddr()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getLocalPort()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public ServletContext getServletContext()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public AsyncContext startAsync() throws IllegalStateException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public AsyncContext startAsync(ServletRequest servletRequest, ServletResponse servletResponse)
		throws IllegalStateException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isAsyncStarted()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isAsyncSupported()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public AsyncContext getAsyncContext()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public DispatcherType getDispatcherType()
	{
		throw new UnsupportedOperationException();
	}
}

/**
 * Converts the servlet API to Jetty's core API.
 */
public final class JettyServletResponse implements HttpServletResponse
{
	private final Response response;

	/**
	 * @param response the server response
	 * @throws NullPointerException if {@code response} is null
	 */
	public JettyServletResponse(Response response)
	{
		requireThat(response, "response").isNotNull();
		this.response = response;
	}

	@Override
	public void addCookie(Cookie cookie)
	{
		HttpCookie httpCookie = from(cookie);
		Response.addCookie(response, httpCookie);
	}

	private static HttpCookie from(Cookie cookie)
	{
		Builder builder = HttpCookie.build(cookie.getName(), cookie.getValue(), cookie.getVersion());
		String comment = cookie.getComment();
		if (comment != null)
			builder.comment(comment);
		String domain = cookie.getDomain();
		if (domain != null)
			builder.domain(domain);
		builder.httpOnly(cookie.isHttpOnly());
		builder.maxAge(Math.clamp(cookie.getMaxAge(), Integer.MIN_VALUE, Integer.MAX_VALUE));
		String path = cookie.getPath();
		if (path != null)
			builder.path(path);
		builder.secure(cookie.getSecure());
		return builder.build();
	}

	@Override
	public boolean containsHeader(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String encodeURL(String url)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String encodeRedirectURL(String url)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	@Deprecated
	public String encodeUrl(String url)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	@Deprecated
	public String encodeRedirectUrl(String url)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void sendError(int sc, String msg) throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void sendError(int sc) throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void sendRedirect(String location) throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setDateHeader(String name, long date)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void addDateHeader(String name, long date)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setHeader(String name, String value)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void addHeader(String name, String value)
	{
		response.getHeaders().add(name, value);
	}

	@Override
	public void setIntHeader(String name, int value)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void addIntHeader(String name, int value)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setStatus(int sc)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	@Deprecated
	public void setStatus(int sc, String sm)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getStatus()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getHeader(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Collection<String> getHeaders(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Collection<String> getHeaderNames()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getCharacterEncoding()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getContentType()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public ServletOutputStream getOutputStream() throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public PrintWriter getWriter() throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setCharacterEncoding(String charset)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setContentLength(int len)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setContentLengthLong(long len)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setContentType(String type)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setBufferSize(int size)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getBufferSize()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void flushBuffer() throws IOException
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void resetBuffer()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isCommitted()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void reset()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setLocale(Locale loc)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Locale getLocale()
	{
		throw new UnsupportedOperationException();
	}
}

/**
 * Converts the servlet API to Jetty's core API.
 */
public final class JettySession implements HttpSession
{
	private final Request request;
	private final Response response;

	/**
	 * @param request  the server request
	 * @param response the server response
	 * @throws NullPointerException if any of the arguments are null
	 */
	public JettySession(Request request, Response response)
	{
		requireThat(request, "request").isNotNull();
		requireThat(response, "response").isNotNull();
		this.request = request;
		this.response = response;
	}

	@Override
	public long getCreationTime()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public String getId()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public long getLastAccessedTime()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public ServletContext getServletContext()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setMaxInactiveInterval(int interval)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public int getMaxInactiveInterval()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	@Deprecated
	public HttpSessionContext getSessionContext()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Object getAttribute(String name)
	{
		Session session = request.getSession(false);
		if (session == null)
			return null;
		return session.getAttribute(name);
	}

	@Override
	@Deprecated
	public Object getValue(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public Enumeration<String> getAttributeNames()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	@Deprecated
	public String[] getValueNames()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void setAttribute(String name, Object value)
	{
		HttpCookie session = getCookieByName(request, "session");
		String attributeValueAsString = (String) value;
		String sessionValue;
		if (session == null)
			sessionValue = "";
		else
			sessionValue = session.getValue();
		Response.putCookie(response, HttpCookie.build("session", sessionValue).
			attribute(name, attributeValueAsString).build());
	}

	/**
	 * Looks up a cookie by its name.
	 *
	 * @param request a request
	 * @param name    the name of a cookie
	 * @return {@code null} if no match was found
	 */
	private static HttpCookie getCookieByName(Request request, String name)
	{
		List<HttpCookie> cookies = Request.getCookies(request);
		if (cookies == null)
			return null;
		for (HttpCookie cookie : cookies)
		{
			if (cookie.getName().equals(name))
				return cookie;
		}
		return null;
	}

	@Override
	@Deprecated
	public void putValue(String name, Object value)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void removeAttribute(String name)
	{
		request.removeAttribute(name);
	}

	@Override
	@Deprecated
	public void removeValue(String name)
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public void invalidate()
	{
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isNew()
	{
		throw new UnsupportedOperationException();
	}
}
```

Do we dump these into the EE environments and implement more methods as users request them? Or do we let this code sit in this ticket for now?

**Comment by ilapavuluri** [comment]
I think the below file causing the issue
https://github.com/jetty/jetty.project/blob/jetty-11.0.x/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java


Appreciate any help on this.

**Comment by gregw** [comment]
@cowwoc actually, we now support cross context dispatch that wraps servlet request as core requests. So we now have the classes needed. We just need to work out of you can use them. Can you do a servlet dispatch to a core context?

**Comment by cowwoc** [comment]
@gregw This use-case came about when I was trying to integrate https://github.com/auth0/auth0-java into a Jetty instance that did not use servlets.

Eventually, I noticed that their library was so broken and their dependencies were extremely out of date, so I dumped their library and just HTTP requests myself. In other words, I am no longer using this bridge.

In terms of what the code looked like with auth0:

Open https://auth0.com/docs/quickstart/webapp/java/01-login and search for this code: `Tokens tokens = authenticationController.handle(req, res);`

Looking at https://github.com/auth0/auth0-java-mvc-common/blob/07fc4ac31d48258cd42bb44e8e56294d7b56919f/src/main/java/com/auth0/AuthenticationController.java#L302 they expect you to pass in a `ServletHttpRequest/Response`.

I hope that helps.
